service/role.go

```diff
@@ -145,109 +215,174 @@ func (r *roleService) getRoleToken(ctx context.Context, domain, role, proxyForPr
+func (r *roleService) updateRoleToken(ctx context.Context, domain, role, proxyForPrincipal string, minExpiry, maxExpiry time.Duration) (*RoleToken, error) {
+	key := encode(domain, role, proxyForPrincipal)
+
+	rt, err, _ := r.group.Do(key, func() (interface{}, error) {
+		rt, e := r.fetchRoleToken(ctx, domain, role, proxyForPrincipal, minExpiry, maxExpiry)
+		if e != nil {
+			return nil, e
 		}
 
-		// set the token into cache
-		r.domainRoleCache.SetWithExpire(encode(domain, role), &cacheData{
-			token:             data,
+		r.domainRoleCache.SetWithExpire(key, &cacheData{
+			token:             rt,
 			domain:            domain,
 			role:              role,
 			proxyForPrincipal: proxyForPrincipal,
 			minExpiry:         minExpiry,
 			maxExpiry:         maxExpiry,
-		}, time.Unix(data.ExpiryTime, 0).Sub(fastime.Now())-time.Minute)
-		return data, nil
+		}, time.Unix(rt.ExpiryTime, 0).Sub(fastime.Now()))
+
+		return rt, nil
 	})
+	if err != nil {
+		return nil, err
+	}
+
+	return rt.(*RoleToken), err
+}
+
+// fetchRoleToken fetch the role token from Athenz server, and return the decoded role token and any error if occurred.
+func (r *roleService) fetchRoleToken(ctx context.Context, domain, role, proxyForPrincipal string, minExpiry, maxExpiry time.Duration) (*RoleToken, error) {
+	glg.Debugf("get role token, domain: %s, role: %s, proxyForPrincipal: %s, minExpiry: %s, maxExpiry: %s", domain, role, proxyForPrincipal, minExpiry, maxExpiry)
+
+	// get the n-token
+	tok, err := r.token()
+	if err != nil {
+		return nil, err
+	}
+
+	// prepare request object
+	u := r.getRoleTokenAthenzURL(domain, role, minExpiry, maxExpiry, proxyForPrincipal)
+	glg.Debugf("url: %v", u)
+	req, err := http.NewRequest(http.MethodGet, u, nil)
+	if err != nil {
+		glg.Debugf("fail to create request object, error: %s", err)
+		return nil, err
+	}
+	req.Header.Set(r.athenzPrincipleHeader, tok)
 
+	res, err := r.httpClient.Do(req.WithContext(ctx))
 	if err != nil {
 		return nil, err
 	}
 
-	return tok.(*RoleToken), nil
+	defer func() {
+		if res != nil {
+			io.Copy(ioutil.Discard, res.Body)
+			if res.Body != nil {
+				res.Body.Close()
+			}
+		}
+	}()
+
+	if res.StatusCode != http.StatusOK {
+		ioutil.ReadAll(res.Body)
+		return nil, ErrRoleTokenRequestFailed
+	}
+
+	var data *RoleToken
+	if err = json.NewDecoder(res.Body).Decode(&data); err != nil {
+		return nil, err
+	}
+
+	return data, nil
 }
 
 func (r *roleService) getCache(domain, role, principal string) (*RoleToken, bool) {
-	val, ok := r.domainRoleCache.Get(encode(domain, role))
+	val, ok := r.domainRoleCache.Get(encode(domain, role, principal))
 	if !ok {
 		return nil, false
 	}
 	return val.(*cacheData).token, ok
 }
 
-func encode(domain, role string) string {
-	return fmt.Sprintf("%s-%s", domain, role)
+func encode(domain, role, principal string) string {
+	s := []string{domain, role, principal}
+	if principal == "" {
```

Comment:
no need for this checking (not sure)? just encode as `domain,role,`

Reply:
I just added a test case.

Reply:
I mean we can remove this `if` checking?
just encode as `domain;role;` to be simple?

Reply:
same feeling as me, but I discussed with nishitani-san and we agreed to make it like this.

Reply:
OK with that.

another thing, I think we need the role sorting logic.
https://github.com/yahoo/athenz/blob/12aefb788a92d4de44f8b4b78cefff648b16ae1c/clients/java/zts/core/src/main/java/com/yahoo/athenz/zts/ZTSClient.java#L1847-L1859
the roles is a set in the server
https://github.com/yahoo/athenz/blob/master/servers/zts/src/main/java/com/yahoo/athenz/zts/ZTSImpl.java#L1357

Reply:
@kevindiu ping

Reply:
Do we need this logic?

Reply:
better to have it?

If user request with "r1,r2" and "r2,r1", we will have 2 caches of the role token

Reply:
fixed! thx!